clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java

```diff
@@ -227,4 +301,130 @@ private KeyStore load() throws GeneralSecurityException, IOException {
+            }
+        }
+
+        private SSLConfigValidatorEngine(SslFactory sslFactory, SSLContext sslContext, Mode mode) {
+            this.sslEngine = sslFactory.createSslEngine(sslContext, "localhost", 0); // these hints are not used for validation
+            sslEngine.setUseClientMode(mode == Mode.CLIENT);
+            appBuffer = ByteBuffer.allocate(sslEngine.getSession().getApplicationBufferSize());
+            netBuffer = ByteBuffer.allocate(sslEngine.getSession().getPacketBufferSize());
+        }
+
+        void beginHandshake() throws SSLException {
+            sslEngine.beginHandshake();
+        }
+
+        void handshake(SSLConfigValidatorEngine peerEngine) throws SSLException {
+            SSLEngineResult.HandshakeStatus handshakeStatus = sslEngine.getHandshakeStatus();
+            while (true) {
+                switch (handshakeStatus) {
+                    case NEED_WRAP:
+                        handshakeResult = sslEngine.wrap(EMPTY_BUF, netBuffer);
+                        switch (handshakeResult.getStatus()) {
+                            case OK: break;
+                            case BUFFER_OVERFLOW:
+                                netBuffer.compact();
+                                Utils.ensureCapacity(netBuffer, sslEngine.getSession().getPacketBufferSize());
+                                netBuffer.flip();
+                                break;
+                            case BUFFER_UNDERFLOW: throw new SSLException("Unexpected status: CLOSED");
+                            case CLOSED: throw new SSLException("Unexpected status: CLOSED");
+                        }
+                        return;
+                    case NEED_UNWRAP:
+                        if (peerEngine.netBuffer.position() == 0)
+                            return;
+                        peerEngine.netBuffer.flip();
+                        handshakeResult = sslEngine.unwrap(peerEngine.netBuffer, appBuffer);
+                        peerEngine.netBuffer.compact();
+                        handshakeStatus = handshakeResult.getHandshakeStatus();
+                        switch (handshakeResult.getStatus()) {
+                            case OK: break;
+                            case BUFFER_OVERFLOW:
+                                Utils.ensureCapacity(appBuffer, sslEngine.getSession().getApplicationBufferSize());
+                                break;
+                            case BUFFER_UNDERFLOW:
+                                Utils.ensureCapacity(netBuffer, sslEngine.getSession().getPacketBufferSize());
+                                break;
+                            case CLOSED: throw new SSLException("Unexpected status: CLOSED");
+                        }
+                        break;
+                    case NEED_TASK:
+                        sslEngine.getDelegatedTask().run();
+                        handshakeStatus = sslEngine.getHandshakeStatus();
```

Comment:
As far as I can tell, every time we iterate the loop, we update `handshakeStatus`. Would it make sense to move its initialization to the first line in the loop and remove these updates?

Reply:
For the `unwrap` case, we get the result from the handshake result of unwrap, so it is different from delegated tasks where the latest status is from the sslEngine. The logic here is the same as `SslTransportLayer` and that contains a lot more comments. Unfortunately it would be too messy to combine the two, so I tried to make this as small as possible.